File: src/scene/credits.rs
use macroquad::color::{RED, WHITE};

use super::Scene;
use crate::audio::play_sfx;
use crate::consts::TITLE_Y_INSET;
use crate::consts::VIRTUAL_HEIGHT;
use crate::consts::X_INSET;
use crate::input::action_pressed;
use crate::input::Action;
use crate::text::Size;
use crate::{context::Context, text::draw_text};

/// sub-scene for displaying who worked on the game
pub struct Credits {
    pub active: bool,
}
impl Credits {
    pub fn new(_ctx: &Context) -> Self {
        Self { active: false }
    }
}

impl Scene for Credits {
    fn update(&mut self, ctx: &mut Context) {
        if action_pressed(Action::Cancel, &ctx.gamepads)
            || action_pressed(Action::Confirm, &ctx.gamepads)
        {
            play_sfx(ctx, &ctx.audio.sfx.menu_cancel);
            self.active = false;
        }
    }

    fn draw(&mut self, ctx: &mut Context) {
        draw_text(ctx, "Credits", X_INSET, TITLE_Y_INSET, Size::Large, WHITE);

        draw_text(
            ctx,
            "Art by Chrysalis, Vellidragon, kenney, and Brett Chalupa",
            X_INSET,
            240.,
            Size::Small,
            WHITE,
        );
        draw_text(
            ctx,
            "Sound effects by Melos Han-Tani",
            X_INSET,
            300.,
            Size::Small,
            WHITE,
        );
        draw_text(
            ctx,
            "Level design by Brett Chalupa and Yoshio Murase",
            X_INSET,
            360.,
            Size::Small,
            WHITE,
        );
        draw_text(
            ctx,
            "Programming by Brett Chalupa",
            X_INSET,
            420.,
            Size::Small,
            WHITE,
        );

        draw_text(
            ctx,
            "Press Z or X to return",
            X_INSET,
            VIRTUAL_HEIGHT - 120.,
            Size::Medium,
            RED,
        );
    }
}

File: scenes/credits.lua
-- Static credits screen. BTN1 or BTN2 returns to the main menu.

local consts = require("consts")

local credits = {}

function credits.enter() end

function credits.leave() end

function credits.update(_dt)
  if input.pressed(input.BTN1) or input.pressed(input.BTN2) then
    sfx.play(consts.SFX.MENU_CANCEL)
    SwitchScene("main_menu")
  end
end

local LINES = {
  "Art by Chrysalis and Brett Chalupa",
  "Sound effects by Melos Han-Tani",
  "Level design by Brett Chalupa and Yoshio Murase",
  "Programming by Brett Chalupa",
}

function credits.draw(_dt)
  gfx.clear(gfx.COLOR_BLACK)
  gfx.text("Credits", 8, 8, gfx.COLOR_WHITE)
  for i, line in ipairs(LINES) do
    gfx.text(line, 8, 28 + (i - 1) * 12, gfx.COLOR_LIGHT_GRAY)
  end
  local back = "Press " .. input.mapping_for(input.BTN1) ..
      " or " .. input.mapping_for(input.BTN2) .. " to return"
  local _, bh = usagi.measure_text(back)
  gfx.text(back, 8, usagi.GAME_H - bh - 2, gfx.COLOR_RED)
end

return credits
